tests/src/test/java/alluxio/master/journal/ufs/UfsJournalLogWriterTest.java

```diff
@@ -178,6 +197,229 @@ public void writeJournalEntryRotate() throws Exception {
     }
   }
 
+  /**
+   * Tests that (1) {@link UfsJournalLogWriter#mJournalOutputStream} is reset when an exception
+   * is thrown, and (2) the {@link UfsJournalLogWriter} recovers during the next write. It should
+   * write out all entries, including the one that initially failed.
+   */
+  @Test
+  public void recoverFromUfsFailure() throws Exception {
+    long startSN = 0x10;
+    UfsJournalLogWriter writer = new UfsJournalLogWriter(mJournal, startSN);
+    final int numberOfEntriesWrittenBeforeFailure = 10;
+    long nextSN = writeJournalEntries(writer, startSN, numberOfEntriesWrittenBeforeFailure);
+    // Create a mock DataOutputStream.
+    DataOutputStream badOut = createMockDataOutputStream(writer);
+    // Specify the behavior of badOut so that it throws an IOException containing
+    // "injected I/O error" when its `write` method is invoked with any arguments
+    // matching (byte[], int, int).
+    Mockito.doThrow(new IOException(INJECTED_IO_ERROR_MESSAGE)).when(badOut)
+        .write(Mockito.any(byte[].class), Mockito.anyInt(), Mockito.anyInt());
+    tryWriteAndExpectToFail(writer, nextSN);
+    // Write another entry. UfsJournalLogWriter will perform recovery first.
```

Comment:
Remove "Write another entry" since that's what the code says. It's fine to comment on why we are calling `write` if it is subtle

Reply:
Done.